les/fetcher.go

```diff
@@ -770,33 +773,39 @@ func (f *lightFetcher) checkSyncedHeaders(p *peer) (bool, []*types.Header, *big.
 	return true, []*types.Header{header}, td
 }
 
-// lastValidTrieNode return last approved treeNode and a list of unapproved hashes
-func (f *lightFetcher) lastValidTrieNode(p *peer) (*types.Header, []common.Hash) {
+// lastTrustedTreeNode return last approved treeNode and a list of unapproved hashes
+func (f *lightFetcher) lastTrustedTreeNode(p *peer) (*types.Header, []common.Hash) {
 	unapprovedHashes := make([]common.Hash, 0)
-
 	current := f.chain.CurrentHeader()
-	for current != nil || current.Hash() != f.getLastValidHeader().Hash() {
+
+	if f.lastTrustedHeader == nil {
+		return current, unapprovedHashes
+	}
+
+	canonical := f.chain.CurrentHeader()
+	if canonical.Number.Uint64() > f.lastTrustedHeader.Number.Uint64() {
+		canonical = f.chain.GetHeaderByNumber(f.lastTrustedHeader.Number.Uint64())
+	}
+	ancestorHash := rawdb.FindCommonAncestor(f.pm.chainDb, canonical, f.lastTrustedHeader).Hash()
```

Comment:
I did not write this in my previous review but I'd add a check to avoid panic if the common ancestor header is nil. In that case print an error log.